### ISSUE TITLE
Add ability to trigger new builds if revision not found in buildapi

### ISSUE THREAD

**Issue by vaibhavmagarwal** [mozilla/mozilla_ci_tools#313]
RyanVM said that he mistakenly pushed with 'DONT BUILD' to treeherder, and tried to trigger missing jobs. However mozci doesn't do anything if it can't find the particular revision. We should trigger new builds in such a case, since we have the ability to trigger_arbitrary_jobs. Relevant logs:
```python
INFO:	 We want to have 1 job(s) of Android 4.0 armv7 API 11+ mozilla-central debug test crashtest on revisions [u'be100e46b1ba'] 
Aug 07 12:48:52 pulse-actions app/worker2.1: INFO:	  
Aug 07 12:48:52 pulse-actions app/worker2.1: INFO:	 === be100e46b1ba === 
Aug 07 12:48:52 pulse-actions app/worker2.1: INFO:	 We can't trigger anything on pushes that the revision is not valid for buildapi. 
Aug 07 12:48:52 pulse-actions app/worker2.1: INFO:	 We want to have 1 job(s) of Rev4 MacOSX Snow Leopard 10.6 mozilla-central opt test mochitest-2 on revisions [u'be100e46b1ba'] 
Aug 07 12:48:52 pulse-actions app/worker2.1: INFO:	  
Aug 07 12:48:52 pulse-actions app/worker2.1: INFO:	 === be100e46b1ba === 
Aug 07 12:48:52 pulse-actions app/worker2.1: INFO:	 We can't trigger anything on pushes that the revision is not valid for buildapi. 
Aug 07 12:48:52 pulse-actions app/worker2.1: INFO:	 We want to have 1 job(s) of Rev4 MacOSX Snow Leopard 10.6 mozilla-central opt test mochitest-3 on revisions [u'be100e46b1ba'] 
Aug 07 12:48:52 pulse-actions app/worker2.1: INFO:	  
Aug 07 12:48:52 pulse-actions app/worker2.1: INFO:	 === be100e46b1ba === 
Aug 07 12:48:52 pulse-actions app/worker2.1: INFO:	 We can't trigger anything on pushes that the revision is not valid for buildapi. 
Aug 07 12:48:52 pulse-actions app/worker2.1: INFO:	 We want to have 1 job(s) of Rev4 MacOSX Snow Leopard 10.6 mozilla-central opt test mochitest-4 on revisions [u'be100e46b1ba'] 
Aug 07 12:48:52 pulse-actions app/worker2.1: INFO:	  
Aug 07 12:48:52 pulse-actions app/worker2.1: INFO:	 === be100e46b1ba === 
Aug 07 12:48:52 pulse-actions app/worker2.1: INFO:	 We can't trigger anything on pushes that the revision is not valid for buildapi. 
Aug 07 12:48:52 pulse-actions app/worker2.1: INFO:	 We want to have 1 job(s) of Rev4 MacOSX Snow Leopard 10.6 mozilla-central opt test mochitest-5 on revisions [u'be100e46b1ba'] 
Aug 07 12:48:52 pulse-actions app/worker2.1: INFO:	  
Aug 07 12:48:52 pulse-actions app/worker2.1: INFO:	 === be100e46b1ba === 
Aug 07 12:48:52 pulse-actions app/worker2.1: INFO:	 We can't trigger anything on pushes that the revision is not valid for buildapi. 
Aug 07 12:48:52 pulse-actions app/worker2.1: INFO:	 We want to have 1 job(s) of Rev4 MacOSX Snow Leopard 10.6 mozilla-central opt test mochitest-browser-chrome-1 on revisions [u'be100e46b1ba'] 
Aug 07 12:48:52 pulse-actions app/worker2.1: INFO:	  
```

**Comment by armenzg** [comment]
This issue is known.
If a push does not schedule anything on the buildbot world, we won't be able to schedule anything with mozci.
Alice brought this to my attention and I believe there is a chance (maybe!) that we can cheat the system and call this other API:
> POST /self-serve/{branch}/rev/{revision} Creates a new build at this revision

If you load a push:
https://secure.pub.build.mozilla.org/buildapi/self-serve/mozilla-central/rev/be100e46b1ba
you can read
" Create new dep builds on mozilla-central revision "

I think that is the API that Ryan triggered.

I believe this is what allowed suddently such revision to have BuildApi support.
We can test this on try by pushing with a try syntax that only triggers a TaskCluster job.

**Comment by armenzg** [comment]
Source code:
http://mxr.mozilla.org/build/source/buildapi/buildapi/templates/self-serve/branch.mako#234

**Comment by vaibhavmagarwal** [comment]
Yes, that is what I was thinking about. We should be able to trigger new builds even if it isn't present on buildapi, if that is possible in the web interface.

**Comment by adusca** [comment]
I tested this endpoint with a revision on try with only TC jobs, and I got an status 'OK', but no new jobs were triggered. I did receive the 'Thank you for your try submission' emails:

request: https://secure.pub.build.mozilla.org/buildapi/self-serve/jobs/1449048
th link: https://treeherder.mozilla.org/#/jobs?repo=try&revision=137f1d4f4114

Doing the same with a job that had buildbot jobs results in the same jobs being pushed again: https://treeherder.mozilla.org/#/jobs?repo=try&revision=8d79187e347f

It looks like this endpoint triggers everything respecting the try commit message, but it doesn't respect DONTBUILD.

**Comment by adusca** [comment]
I decided to test:
```
trigger_arbitrary_job('try', 'Linux try build', '137f1d4f4114')
```

Before doing this, https://secure.pub.build.mozilla.org/buildapi/self-serve/try/rev/137f1d4f4114 returned a 404, now it shows a list of jobs. That means we can call trigger_arbitrary_job for revisions that do not exist in buildapi!

**Comment by vaibhavmagarwal** [comment]
Yeah, it makes sense that trigger_arbitrary_job is able to trigger new builds. So currently we already do this, but we check if that revision is already valid for buildapi. I guess we should tweak the validation logic and trigger builds in case we do not find it.

**Comment by vaibhavmagarwal** [comment]
@adusca I tested your function new_build_for_revision on mozilla-central and it worked for me. See
https://treeherder.mozilla.org/#/jobs?repo=mozilla-central&revision=4499640fe85c

However, it didn't work for me on try.

**Comment by armenzg** [comment]
I wonder why I ever validated if the revision existed.
Maybe I tried to query jobs from BuildAPI when there was no valid revision?

In any case thanks for fixing this!

**Comment by adusca** [comment]
Fixed in #315
